fix: Ensure focus changes when tabbing fields

BenHenning opened the pull request:
## The basics

- [x] I [validated my changes](https://developers.google.com/blockly/guides/contribute/core#making_and_verifying_a_change)

## The details
### Resolves

Fixes https://github.com/google/blockly-keyboard-experimentation/issues/578
Fixes part of #8915

### Proposed Changes

Ensures fields update focus to the next field when tabbing between field editors. The old behavior can be observed in [#578](https://github.com/google/blockly-keyboard-experimentation/issues/578) and the new behavior can be observed here:

[Screen recording 2025-06-25 1.39.28 PM.webm](https://github.com/user-attachments/assets/e00fcb55-5c20-4d5c-81a8-be9049cc0d7e)

### Reason for Changes

Having focus reset back to the original field editor that was opened is an unexpected experience for users. This approach is better.

Note that there are some separate changes added here, as well:
- Connections and fields now check if their block IDs contain their indicator prefixes since this will all-but-guarantee focus breaks for those nodes. This is an excellent example of why #9171 is needed.
- Some minor naming updates for `FieldInput`: it was incorrectly implying that key events are sent for changes to the `input` element used by the field editor, but they are actually `InputEvent`s per https://developer.mozilla.org/en-US/docs/Web/API/Element/input_event.

### Test Coverage

New tests were added for field editing in general (since this seems to be missing), including tabbing support to ensure the fixes originally introduced in #9049.

One new test has been added specifically for verifying that focus updates with tabbing. This has been verified to fail with the fix removed (as have all tabbing tests with the tabbing code from #9049 removed).

Some specific notes for the test changes:
- There's a slight test dependency inversion happening here. `FieldInput` is being tested with a specific `FieldNumber` class via real block loading. This isn't ideal, but it seems fine given the circumstances (otherwise a lot of extra setup would be necessary for the tests).
- The workspace actually needs to be made visible during tests in order for focus to work correctly (though it's reset at the end of each test, but this may cause some flickering while the tests are running).
- It's the case that a bunch of tests were actually setting up blocks incorrectly (i.e. not defining a must-have `id` property which caused some issues with the new field and connection ID validation checks). These tests have been corrected, but it's worth noting that the blocks are likely still technically wrong since they are not conforming to their TypeScript contracts.
- Similar to the previous point, one test was incorrectly setting the first ID to be returned by the ID generator as `undefined` since (presumably due to a copy-and-paste error when the test was introduced) it was referencing a `TEST_BLOCK_ID` property that hadn't been defined for that test suite. This has been corrected as, without it, there are failures due to the new validation checks.
- For the connection database checks, a new ID is generated instead of fixing the block ID to ensure that it's always unique even if called multiple times (otherwise a block ID would need to be piped through from the calling tests, or an invalid situation would need to be introduced where multiple blocks shared an ID; the former seemed unnecessary and the latter seemed nonideal).
- There are distinct Geras/Zelos tests to validate the case where a full-block field should have its parent block, rather than the field itself, focused on tabbing. See this conversation for more context: https://github.com/google/blockly/pull/9173#discussion_r2172921455.

### Documentation

No documentation changes should be needed here.

### Additional Information

Nothing to add.

BenHenning committed:
fix: Ensure focus changes when tabbing fields.

BenHenning reviewed (commented):
Spot checked changes.

BenHenning committed:
fix: Other broken tests.

These were setting up blocks or, in one case, the id generator
incorrectly.

BenHenning reviewed (commented):
Spot-checked test fix changes.

gonfunko commented on core/field_input.ts:
```diff
@@ -614,6 +615,7 @@ export abstract class FieldInput<T extends InputTypes> extends Field<
       if (target instanceof FieldInput) {
         WidgetDiv.hideIfOwner(this);
         dropDownDiv.hideWithoutAnimation();
+        getFocusManager().focusNode(target);
```
Drive by comment: In the case of full-block fields, I think we actually want to focus the parent block for consistency.

cpcallen replied:
Yes, agreed.  (And note that whether a field is a full-block field or not depends on the renderer.)

BenHenning replied:
Didn't know about full-block fields before this. Fortunately, the block being tested here already uses full-block fields in Zelos, so I updated the tests to verify behavior specifically with Zelos and Geras separately. I confirmed that one of the Zelos tests fails without the new `focusNode` line in place.

cpcallen commented on core/connection.ts:
```diff
@@ -83,6 +83,12 @@ export class Connection {
+        `Connection ID indicator is contained in block ID. This may cause ` +
+          `problems with focus: ${source.id}.`,
+      );
+    }
```
Maybe I am misunderstanding something, but is it not an error for a block to have an ID containing `'_connection'`?  And if it is an error, should this not be `throw new TypeError` (or at least `console.error`) instead of merely `console.warn`?

Also, are we not testing for this in the `Block` constructor?

BenHenning replied:
I don't see any specific validation for ID in `Block` or `BlockSvg`'s constructors, but maybe there's somewhere subtle where it happens? We could check for the field & connection failure cases there, instead.

It probably is an error, I was just hesitant to make it actually throw. I'm fine with doing that, however. Updated. That being said, I'm using `Error`, not `TypeError`. The type (string) is technically correct which makes this a value, not type, incompatibility.

I'm less certain about moving the checks elsewhere because these failure cases actually came up from weirdly set up tests. We have a BUNCH of tests that set up invalid blocks by defining them directly as JSON objects that don't actually define required properties like `id`. I fixed the ones that were relevant here, but that does mean that the constructor isn't guaranteed to be called (and thus this check could miss real failures in tests even if it works correctly for production code).

cpcallen replied:
If I understand correctly #9171 should address the problem that these checks are trying to guard against, but if that's not the case would you open an issue to ensure that we have suitable checks (in constructor, deserialiser, etc.) to ensure that invariants about identifiers are maintained?

BenHenning replied:
That will fix it. Do you still think the checks are needed in the meantime?

Will follow up separately so that I can merge this PR.

cpcallen replied:
Further checks are not needed for this PR.

I will leave it to you to decide if there is cause to add other interim checks before the matter can be addressed by #9171.

BenHenning replied:
In that case, I'll probably hold off but I'll make note of this conversation in that thread.

cpcallen commented on core/field_input.ts:
```diff
@@ -83,8 +84,8 @@ export abstract class FieldInput<T extends InputTypes> extends Field<
-  /** Key input event data. */
-  private onKeyInputWrapper: browserEvents.Data | null = null;
+  /** Input element input event data. */
+  private onInputWrapper: browserEvents.Data | null = null;
```
Since you're working on improving naming:

I don't think a variable of type `Data` should be called `wrapper`, since the latter strongly implies (to me at least) that the value of the variable will be a callable function—which AFAICT this is not.  Maybe:

```suggestion
  private onInputBinding: browserEvents.Data | null = null;
```

BenHenning replied:
Did a quick audit and it seems we call the returned object any of the following:
- Wrapper (here and three other files)
- Event (one file)
- Listener (three files)
- Handler (two files)

I'm not opposed to fixing the naming, but this introduces a new type and thus technically increases the inconsistency. I can rename all of the cases where it's used, but I lean toward doing that in a follow-up PR to keep it separate (in case this needs to be reverted for some reason).

Does that sound good to you? If so, do you have a strong opinion on 'binding' being the "one true name suffix" we should use everywhere? (I think 'binding' is reasonable, for what it's worth, though I might lean a bit more toward 'eventBinding' to be extra clear).

cpcallen replied:
I agree consistency is good, but am happy for you to leave it to a subsequent PR.

I have no strong preference about the exact naming, though given my background in C and Go I generally prefer shorter names than some of our erstwhile Java-programming colleagues.

cpcallen commented on tests/mocha/field_textinput_test.js:
```diff
@@ -23,11 +23,50 @@ import {
 
 suite('Text Input Fields', function () {
-  setup(function () {
+  setup(async function () {
```
I don't think this needs to be an async function: its body contains no occurrences of `await`.

BenHenning replied:
I think this is a remnant from when I previously had a call to `finishQueuedRenders` here. Nice catch--removed.

BenHenning committed:
Merge branch 'develop' into focus-when-tabbing-fields

BenHenning committed:
fix: Make full-block field tabbing work.

This includes adding new tests.

Also separately: replace the connection and field ID check warnings with
actual thrown errors.

All of this addresses a number of reviewer comments.

BenHenning committed:
Merge branch 'develop' into focus-when-tabbing-fields

BenHenning reviewed (commented):
Spot-checked latest changes.
